core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
   }
 
   /**
-   * Get the transaction metadata associated with the given transactional id, or null if not found
+   * Get the transaction metadata associated with the given transactional id, or None if not found, or an error if
+   * the coordinator does not own the transaction partition or is still loading it
+   *
+   * Need to grab the state lock before calling this function
    */
-  def getTransactionState(transactionalId: String): Option[CoordinatorEpochAndTxnMetadata] = {
+  def getTransactionState(transactionalId: String): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]] = inReadLock(stateLock) {
     val partitionId = partitionFor(transactionalId)
 
-    // we only need to check leaving partition set but not loading partition set since there are three possible cases:
-    //    1) it is not in the loading partitions set, hence safe to return NONE
-    //    2) it is in the loading partitions with a smaller epoch, hence safe to return NONE
-    //    3) it is in the loading partition with a larger epoch, return NONE is also fine as it
-    //       indicates the metadata is not exist at least for now.
+    // we only need to check loading partition set but not leaving partition set here since there are three possible cases:
+    //    1) it is not in the leaving partitions set, hence safe to return COORDINATOR_LOAD_IN_PROGRESS
+    //    2) it is in the leaving partitions with a smaller epoch than the latest loading epoch, hence safe to return COORDINATOR_LOAD_IN_PROGRESS
+    //    3) it is in the leaving partition with a larger epoch, return true is also OK since the client will then retry
+    //       later be notified that this coordinator is no longer be the transaction coordinator for him
     //
-    //    4) it is NOT possible to be in the loading partition with the same epoch
-    if (leavingPartitions.exists(_.txnPartitionId == partitionId))
-      return None
-
-    transactionMetadataCache.get(partitionId).flatMap { cacheEntry =>
-      cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
-        case null => None
-        case txnMetadata => Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata))
+    //    4) it is NOT possible to be in the leaving partition with the same epoch
+    if (loadingPartitions.exists(_.txnPartitionId == partitionId))
+      return Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
+
+    transactionMetadataCache.get(partitionId) match {
+      case Some(cacheEntry) =>
+        cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
+          case null =>
+            Right(None)
+          case txnMetadata =>
+            // we only need to check leaving partition set but not loading partition set here since there are three possible cases:
```

Comment:
Is this comment still needed? We already checked the loading partition set above and we are holding the lock.